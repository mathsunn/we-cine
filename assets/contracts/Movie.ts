import { Video } from "./Video";

export interface Movie {
  id: number;
  originalTitle: string;
  title: string;
  overview: string;
  releaseDate: string;
  voteAverage: number;
  voteCount: number;
  videos: Record<string, Video>;
}
