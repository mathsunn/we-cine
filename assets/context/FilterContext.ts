import { createContext, useContext } from "react";
import { Genre } from "../contracts/Genre";

export type FilterContextType = {
  selectedGenre: Genre | null;
  selectedTitle: string | null;
  setGenre: (genre: Genre) => void;
  setTitle: (title: string) => void;
};

export const FilterContext = createContext<FilterContextType>({
  selectedGenre: null,
  selectedTitle: null,
  setGenre: (setGenre) => setGenre,
  setTitle: (setTile) => setTile,
});
export const useFilters = (): FilterContextType => useContext(FilterContext);
