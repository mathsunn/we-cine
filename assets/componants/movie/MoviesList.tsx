import React, { useState } from "react";
import { useFilters } from "../../context/FilterContext";
import useMovies from "../../hooks/UseMovies";
import { isEmpty, map } from "lodash";
import Star from "../star/Star";
import MovieModal from "./MovieModal";

const MoviesList: React.FC = () => {
  const { selectedGenre, selectedTitle } = useFilters();
  const { movies, total, loading } = useMovies(selectedGenre);
  const [currentMovie, setCurrentMovie] = useState(null);

  if (loading) {
    return <>Loading</>;
  }

  if (total === 0) {
    return <>No result</>;
  }

  let filteredMovies = { ...movies };
  if (selectedTitle) {
    filteredMovies = Object.keys(movies)
      .filter(function (key) {
        return (
          null !== movies[key].title.match(new RegExp(selectedTitle, "gi"))
        );
      })
      .reduce((res, key) => ((res[key] = movies[key]), res), {});
  }

  return (
    <>
      {currentMovie && (
        <MovieModal
          currentMovie={currentMovie}
          setCurrentMovie={setCurrentMovie}
        />
      )}
      {isEmpty(filteredMovies) && <p>Movie not found</p>}

      {map(filteredMovies, (movie) => (
        <div className="card" key={movie.id} id={`movie_${movie.id}`}>
          <div className="card-body">
            <h5 className="card-title">
              {movie.title}
              <span>
                <Star count={5} average={Math.ceil(movie.voteAverage / 2)} />
              </span>
              <span>({movie.voteCount} votes)</span>
            </h5>
            <h6 className="card-subtitle mb-2 text-muted">
              {new Date(movie.releaseDate).getFullYear()}
            </h6>
            <p className="card-text">{movie.overview}</p>
            <button
              className="card-link btn btn-primary"
              onClick={() => {
                setCurrentMovie(movie);
              }}
            >
              Open
            </button>
          </div>
        </div>
      ))}
    </>
  );
};

export default MoviesList;
