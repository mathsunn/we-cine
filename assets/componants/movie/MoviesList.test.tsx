import React from "react";
import { render } from "@testing-library/react";
import axios from "axios";
import MoviesList from "./MoviesList";
import { FilterContext } from "../../context/FilterContext";
import { Genre } from "../../contracts/Genre";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

test("matches the snapshot", async () => {
  mockedAxios.get.mockResolvedValue({
    data: {
      "00000000670369ed0000000036c97f7e": {
        id: 1,
        originalTitle: "originalTitle",
        title: "title",
        overview: "overview",
        releaseDate: "2021-01-12T00:00:00+00:00",
        voteAverage: 8.1,
        voteCount: 100,
        videos: [],
      },
    },
  });

  let selectedGenre: Genre = {
    id: 12,
    name: "Action",
  };

  let selectedTitle = null;
  const setGenre = (setGenre) => setGenre;
  const setTitle = (setTile) => setTile;

  const { asFragment, findByText } = render(
    <FilterContext.Provider
      value={{ selectedGenre, setGenre, selectedTitle, setTitle }}
    >
      <MoviesList />
    </FilterContext.Provider>
  );

  await findByText(/title/i);

  expect(asFragment()).toMatchSnapshot();
});
