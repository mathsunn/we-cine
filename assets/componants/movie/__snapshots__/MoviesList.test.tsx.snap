// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="card"
    id="movie_1"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title"
      >
        title
        <span>
          <ul
            class="ratings"
          >
            <li
              class="star selected"
            />
            <li
              class="star selected"
            />
            <li
              class="star selected"
            />
            <li
              class="star selected"
            />
            <li
              class="star selected"
            />
          </ul>
        </span>
        <span>
          (100 votes)
        </span>
      </h5>
      <h6
        class="card-subtitle mb-2 text-muted"
      >
        2021
      </h6>
      <p
        class="card-text"
      >
        overview
      </p>
      <button
        class="card-link btn btn-primary"
      >
        Lire la suite
      </button>
    </div>
  </div>
</DocumentFragment>
`;
